### PR TITLE
Move horizon planning to operators

### PR DESCRIPTION
**Pull request by frouioui**
## Description
A refactoring that introduces a Horizon as a logical operator.

During query planning, we start by figuring out where we are going to send the queries. This is done by inspecting the `FROM` & `WHERE` clauses. 

A horizon is what we call in the planner the stuff that is outside the `FROM ... WHERE` clause. If we are able to send the whole query to a single shard, we don't do much more planning - we simply add the rest of the clauses (projections, ordering, limit, aggregation and more) to the query and we are done.

If we can't delegate everything to MySQL, we do more planning, and this is what we call horizon planning.

This change moves the horizon into the operator tree, instead of having special handling of derived tables, unions and subqueries.

## Related Issue(s)
#11626

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func NewApplyJoin(lhs, rhs Operator, predicate sqlparser.Expr, leftOuterJoin boo
 func (a *ApplyJoin) IPhysical() {}
 
 // Clone implements the Operator interface
-func (a *ApplyJoin) Clone(inputs []Operator) Operator {
+func (a *ApplyJoin) clone(inputs []Operator) Operator {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:71]: nit: we can also change the comment `Clone` -> `clone`

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+func planHorizon(ctx *plancontext.PlanningContext, in *Horizon) (Operator, error) {
+	rb, isRoute := in.Source.(*Route)
+	if isRoute && rb.IsSingleShard() {
+		return planSingleShardRoute(in.Select, rb, in)
+	}
+
+	return nil, errNotHorizonPlanned
+}
+func planSingleShardRoute(statement sqlparser.SelectStatement, rb *Route, horizon *Horizon) (Operator, error) {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:53]: nit: new line :)

File: go/vt/vtgate/planbuilder/operators/horizon.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+*/
+
+package operators
+
+import (
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
+)
+
+// Horizon is an operator we use until we decide how to handle the source to the horizon.
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon.go:24]: I am not sure I fully understand this. We add the `Horizon` operator into the tree until we have decided how the rest of the tree will do horizon planning?

Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/horizon.go:24]: Looking at the operator test data, it feels like it encapsulates each node in the tree.

File: go/vt/vtgate/planbuilder/operators/operator.go [hunk 4]
```diff
@@ -14,13 +14,31 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// Package operators contains the operators used to plan queries.
+/*
+The operators go through a few phases while planning:
+1. Logical
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/operator.go:20]: That's a nice write up

File: go/vt/vtgate/planbuilder/operators/horizon.go (hunk 3 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/horizon.go:24]: Yes. We need to do the physical route planning before we do the horizon planning

File: go/vt/vtgate/planbuilder/operators/operator.go (hunk 4 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/operator.go:20]: I'll add info about horizons in the op tree as well

**Review by frouioui — COMMENTED** [review]
LGTM, can't approve as this is my PR.

**Comment by frouioui** [comment]
We should update the Pull Request description.

**Review by GuptaManan100 — APPROVED** [review]
I like the changes!

**Pull request merged**